dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -433,6 +445,48 @@ private String channelName() {
         }
     }
 
+    private static final class ConsumptionTrackingRequestBody implements RequestBody {
+        private final RequestBody delegate;
+        private volatile boolean consumed;
+
+        ConsumptionTrackingRequestBody(RequestBody delegate) {
+            this.delegate = delegate;
+            this.consumed = false;
+        }
+
+        @Override
+        public void writeTo(OutputStream output) throws IOException {
+            this.consumed = true;
+            delegate.writeTo(output);
+        }
+
+        @Override
+        public String contentType() {
+            return delegate.contentType();
+        }
+
+        @Override
+        public boolean repeatable() {
+            return delegate.repeatable();
+        }
+
+        @Override
+        public void close() {
+            // Closing the body returns all resources, and the body can be considered consumed.
+            this.consumed = true;
+            delegate.close();
+        }
+
+        public boolean requestBodyCanBeRetried() {
```

Comment:
No need for a public modifier on a private class -- it's required on the `@Override` methods because interface method overrides must always be public.
```suggestion
       boolean requestBodyCanBeRetried() {
```